File: apple/WordPressTVCore/Sources/WordPressTVCore/Data/WPComContentRepository.swift
import Foundation

/// `ContentRepository` backed by the WP.com REST API (v1.1).
///
/// For a public source (wordpress.tv, `auth == .none`) it sends no token, just
/// like before. For a private source (a8c.tv, `auth == .wpcomOAuth`) it asks the
/// injected `AuthTokenProviding` for the user's token and sets `Authorization:
/// Bearer`. When the source `needsPlaybackToken`, the poster and stream URLs get
/// a VideoPress `metadata_token` appended — reusing the per-video token WP.com
/// mints into the post's private embed (`Video.playbackToken`) rather than
/// minting one (that needs the broad `global` scope the narrow token lacks). A
/// 401/403 surfaces as `RepositoryError.unauthorized` so the UI can re-pair.
public final class WPComContentRepository: ContentRepository {
    private static let apiBase = URL(string: "https://public-api.wordpress.com/rest/v1.1")!

    private let session: URLSession
    private let decoder = JSONDecoder()
    /// Page size for `listLatest` (the `number` query param).
    private let pageSize: Int
    /// Supplies the Bearer token for authenticated sources; `nil` for a
    /// public-only build.
    private let authProvider: AuthTokenProviding?

    public init(
        session: URLSession = .shared,
        pageSize: Int = 24,
        authProvider: AuthTokenProviding? = nil
    ) {
        self.session = session
        self.pageSize = pageSize
        self.authProvider = authProvider
    }

    // MARK: Implemented

    public func listLatest(source: ContentSource, page: Int) async throws -> [Video] {
        guard var components = URLComponents(
            url: Self.apiBase.appending(path: "sites/\(source.wpcomSite)/posts"),
            resolvingAgainstBaseURL: false
        ) else { throw RepositoryError.invalidURL }

        components.queryItems = [
            URLQueryItem(name: "number", value: String(pageSize)),
            URLQueryItem(name: "page", value: String(max(1, page))),
            URLQueryItem(name: "order_by", value: "date"),
        ]
        guard let url = components.url else { throw RepositoryError.invalidURL }

        let dto: PostsResponseDTO = try await get(url, token: await token(for: source))
        return Mapping.videos(from: dto, sourceID: source.id)
    }

    public func resolvePlayback(source: ContentSource, video: Video) async throws -> PlaybackAsset {
        let token = await token(for: source)
        let url = Self.apiBase.appending(path: "videos/\(video.videoGuid)")
        let dto: VideoInfoDTO = try await get(url, token: token)
        // Private VideoPress (a8c.tv) plays the progressive `original` MP4 with a
        // metadata token appended; public videos (wordpress.tv) keep HLS.
        guard let asset = Mapping.playbackAsset(
            from: dto,
            fallbackTitle: video.title,
            preferProgressive: source.needsPlaybackToken
        ) else {
            throw RepositoryError.notPlayable
        }

        guard source.needsPlaybackToken else { return asset }
        // Private VideoPress (a8c.tv) plays the progressive `original` MP4 with the
        // VideoPress metadata token appended:  "<original>?metadata_token=<token>".
        // Reuse the token WP.com minted into the post's embed (carried on the
        // Video) — minting our own needs the broad `global` OAuth scope. A post
        // without a token is a public video; play it bare rather than treating
        // the absence as an auth failure (which would force a needless re-pair).
        guard let playbackToken = video.playbackToken else { return asset }
        return asset.appendingQuery(name: "metadata_token", value: playbackToken)
    }

    public func posterURL(source: ContentSource, video: Video) async -> URL? {
        // Public sources (wordpress.tv): the posts list already carried a poster
        // from the video's attachment thumbnails — serve it openly.
        guard source.needsPlaybackToken else { return video.posterUrl }
        // Private VideoPress (a8c.tv): posts carry no attachment, so resolve the
        // real poster from the video-info endpoint (the user's `videos` scope
        // authorizes it). Append the embed's metadata token when the post had one
        // — a private poster on videos.files.wordpress.com is 403 without it; a
        // public video's poster serves either way. Best-effort: nil (placeholder)
        // on any failure.
        guard let token = await token(for: source) else { return video.posterUrl }
        let infoURL = Self.apiBase.appending(path: "videos/\(video.videoGuid)")
        do {
            let info: VideoInfoDTO = try await get(infoURL, token: token)
            guard let posterString = info.poster, let poster = URL(string: posterString) else {
                return nil
            }
            guard let playbackToken = video.playbackToken else { return poster }
            return poster.appendingQueryItem(name: "metadata_token", value: playbackToken)
        } catch {
            return nil
        }
    }

    // MARK: Stubbed (later slices)

    public func listCategories(source: ContentSource) async throws -> [CategoryRef] {
        throw RepositoryError.notImplemented
    }

    public func listByCategory(source: ContentSource, category: CategoryRef, page: Int) async throws -> [Video] {
        throw RepositoryError.notImplemented
    }

    public func search(source: ContentSource, query: String, page: Int) async throws -> [Video] {
        throw RepositoryError.notImplemented
    }

    // MARK: Auth helpers

    private func token(for source: ContentSource) async -> String? {
        guard source.auth != .none else { return nil }
        return await authProvider?.accessToken(for: source)
    }

    // MARK: Transport

    private func get<T: Decodable>(_ url: URL, token: String?) async throws -> T {
        var request = URLRequest(url: url)
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return try await send(request)
    }

    private func send<T: Decodable>(_ request: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw RepositoryError.invalidResponse
        }
        if http.statusCode == 401 || http.statusCode == 403 {
            throw RepositoryError.unauthorized
        }
        guard (200..<300).contains(http.statusCode) else {
            throw RepositoryError.http(status: http.statusCode)
        }
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw RepositoryError.decodingFailed
        }
    }
}

private extension PlaybackAsset {
    /// Return a copy with an extra query item on the stream URL (used to append
    /// the VideoPress `metadata_token`).
    func appendingQuery(name: String, value: String) -> PlaybackAsset {
        PlaybackAsset(
            url: url.appendingQueryItem(name: name, value: value),
            kind: kind,
            title: title,
            durationSeconds: durationSeconds
        )
    }
}

extension URL {
    /// Append a single query item, preserving any existing ones.
    func appendingQueryItem(name: String, value: String) -> URL {
        guard var components = URLComponents(url: self, resolvingAgainstBaseURL: false) else { return self }
        components.queryItems = (components.queryItems ?? []) + [URLQueryItem(name: name, value: value)]
        return components.url ?? self
    }
}

File: android/core/src/main/kotlin/com/automattic/wordpresstv/core/data/WpComContentRepository.kt
package com.automattic.wordpresstv.core.data

import com.automattic.wordpresstv.core.domain.CategoryRef
import com.automattic.wordpresstv.core.domain.ContentSource
import com.automattic.wordpresstv.core.domain.PlaybackAsset
import com.automattic.wordpresstv.core.domain.Video
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request

/**
 * [ContentRepository] backed by the WP.com REST API (v1.1).
 *
 * For a public source (wordpress.tv, `auth == NONE`) it sends no token. For a
 * private source (a8c.tv, `auth == WPCOM_OAUTH`) it asks the injected
 * [AuthTokenProvider] for the user's token and sets `Authorization: Bearer`. When
 * the source `needsPlaybackToken`, the poster and stream URLs get a VideoPress
 * `metadata_token` appended — reusing the per-video token WP.com mints into the
 * post's private embed ([Video.playbackToken]). A 401/403 surfaces as
 * [RepositoryException.Unauthorized] so the UI can re-pair.
 */
class WpComContentRepository(
    private val client: OkHttpClient = OkHttpClient(),
    private val pageSize: Int = 24,
    private val authProvider: AuthTokenProvider? = null,
) : ContentRepository {

    private val apiBase = "https://public-api.wordpress.com/rest/v1.1".toHttpUrl()
    private val json = Json { ignoreUnknownKeys = true }

    // --- Implemented ---

    override suspend fun listLatest(source: ContentSource, page: Int): List<Video> {
        val url = apiBase.newBuilder()
            .addPathSegments("sites/${source.wpcomSite}/posts")
            .addQueryParameter("number", pageSize.toString())
            .addQueryParameter("page", maxOf(1, page).toString())
            .addQueryParameter("order_by", "date")
            .build()
        val dto = decode<PostsResponseDto>(getBody(url.toString(), tokenFor(source)))
        return Mapping.videos(dto, source.id)
    }

    override suspend fun resolvePlayback(source: ContentSource, video: Video): PlaybackAsset {
        val token = tokenFor(source)
        val url = apiBase.newBuilder().addPathSegments("videos/${video.videoGuid}").build()
        val dto = decode<VideoInfoDto>(getBody(url.toString(), token))

        // Private VideoPress (a8c.tv) plays the progressive `original` MP4 with a
        // metadata token appended; public videos (wordpress.tv) keep HLS.
        val asset = Mapping.playbackAsset(
            info = dto,
            fallbackTitle = video.title,
            preferProgressive = source.needsPlaybackToken,
        ) ?: throw RepositoryException.NotPlayable

        if (!source.needsPlaybackToken) return asset
        // Reuse the token WP.com minted into the post's embed (carried on the
        // Video). A post without a token is a public video; play it bare rather
        // than treating the absence as an auth failure (a needless re-pair).
        val playbackToken = video.playbackToken ?: return asset
        return asset.copy(url = appendQuery(asset.url, "metadata_token", playbackToken))
    }

    override suspend fun posterUrl(source: ContentSource, video: Video): String? {
        // Public sources (wordpress.tv): the posts list already carried a poster
        // from the video's attachment thumbnails — serve it openly.
        if (!source.needsPlaybackToken) return video.posterUrl

        // Private VideoPress (a8c.tv): posts carry no attachment, so resolve the
        // real poster from the video-info endpoint and append the embed's
        // metadata token. Best-effort: null (placeholder) on any failure.
        val token = tokenFor(source) ?: return video.posterUrl
        val url = apiBase.newBuilder().addPathSegments("videos/${video.videoGuid}").build()
        return try {
            val info = decode<VideoInfoDto>(getBody(url.toString(), token))
            val poster = info.poster ?: return null
            val playbackToken = video.playbackToken ?: return poster
            appendQuery(poster, "metadata_token", playbackToken)
        } catch (_: Exception) {
            null
        }
    }

    // --- Stubbed (later slices) ---

    override suspend fun listCategories(source: ContentSource) =
        throw RepositoryException.NotImplemented

    override suspend fun listByCategory(source: ContentSource, category: CategoryRef, page: Int) =
        throw RepositoryException.NotImplemented

    override suspend fun search(source: ContentSource, query: String, page: Int) =
        throw RepositoryException.NotImplemented

    // --- Auth helpers ---

    private suspend fun tokenFor(source: ContentSource): String? {
        if (source.auth == ContentSource.Auth.NONE) return null
        return authProvider?.accessToken(source)
    }

    // --- Transport ---

    private suspend fun getBody(url: String, token: String?): String =
        withContext(Dispatchers.IO) {
            val builder = Request.Builder().url(url)
            if (token != null) builder.header("Authorization", "Bearer $token")
            client.newCall(builder.build()).execute().use { response ->
                val code = response.code
                if (code == 401 || code == 403) throw RepositoryException.Unauthorized
                if (code !in 200..299) throw RepositoryException.Http(code)
                response.body?.string() ?: throw RepositoryException.InvalidResponse
            }
        }

    private inline fun <reified T> decode(body: String): T =
        try {
            json.decodeFromString<T>(body)
        } catch (_: Exception) {
            throw RepositoryException.DecodingFailed
        }

    /** Append a single query item to [url], preserving any existing ones. */
    private fun appendQuery(url: String, name: String, value: String): String {
        val httpUrl = url.toHttpUrlOrNull() ?: return url
        return httpUrl.newBuilder().addQueryParameter(name, value).build().toString()
    }
}
